### ISSUE TITLE
Native JSON accessor

### ISSUE THREAD

**Issue by techtonik** [ipython/ipython#11553]
I'd like use IPython for quick prototyping with external APIs, and it would be nice if JSON work could be made native without the need to import external libraries. Because JSON is the most common format for API and the web.

**Comment by techtonik** [comment]
If I read data from .json, like:

```
me = """
[
  {
    "id": 43799,
    "name": "anatoly techtonik",
    "username": "techtonik",
    "state": "active",
    "avatar_url": "https://secure.gravatar.com/avatar/9d7e611f31c52f4d62bbe279d4f334de?s=80&d=identicon",
    "web_url": "https://gitlab.com/techtonik"
  }
]
"""
```
Then I'd like to use http://jmespath.org/tutorial.html to access data inside with less boilerplate than `import json<enter>json.loadwhatever(me)[0]["id"]`. 

```
myid = me.json()[0].id
```

**Comment by techtonik** [comment]
`.j` can available as a property of https://ipython.readthedocs.io/en/stable/interactive/shell.html#string-lists if there is a way to turn string into string list easily.

**Comment by takluyver** [comment]
I don't think this is something IPython should do. The focus is on the interface now, not on providing concise APIs for particular operations.

If you want a more convenient way to work with JSON, write a separate module for that. It could provide an [IPython extension](https://ipython.readthedocs.io/en/stable/config/extensions/index.html).

**Comment by techtonik** [comment]
If the focus is on the interface, what is the goals of the interface?

**Comment by takluyver** [comment]
- Providing the default Python kernel for Jupyter applications
- Offering a powerful interactive shell for Python to run in the terminal

A bunch of other stuff got added to IPython in the past, which probably wouldn't be added today. Some of this was general utilities from the days when Python packaging was so awkward that people preferred to make big collections of useful pieces to avoid installing more than one package. Those days are thankfully behind us - Python packaging still isn't great, but it's good enough to distribute much more focused packages.

Another big chunk of functionality from the past is code to make IPython behave more like a shell. Bits of this are useful, but a lot of it is ~unused and rusting, and the code to support it is convoluted and prone to corner cases. [Xonsh](http://xonsh.org/) is a more modern codebase which provides similar features - as it was designed to do this from the beginning, it has a much cleaner approach than IPython can use.

**Comment by techtonik** [comment]
I will take a look at `xonsh`.

So you're telling that the goal right now is to ship Python kernel for Jupiter and all helpers that are implemented so far is a mistake? What is the details of criteria for "powerful interactive shell"? I would benefit from autoimports like in golang in this shell, so I won't have to bail out to mess with `requirements.txt`.

**Comment by takluyver** [comment]
As for your specific proposal: I don't think it needs to be part of IPython, but something like this:

```python
from json import loads

class attr_access(dict): 
     def __getattr__(self, name): 
         if name in self: 
             return self[name] 
         raise AttributeError(name) 

def json(s):
    return loads(s, object_hook=attr_access
```

Would let you do almost what you suggested:

```python
myid = json(me)[0].id
```

If an import is an annoyance, you can use a [startup file](https://ipython.readthedocs.io/en/stable/interactive/tutorial.html#startup-files) to do imports every time you start IPython.

**Comment by techtonik** [comment]
> Xonsh is a more modern codebase which provides similar features - as it was designed to do this from the beginning, it has a much cleaner approach than IPython can use.

I tried `xonsh`. The major design flaw is that separation between shell commands and Python are context dependent. 
```
>>> ls -l
total 0
-rw-rw-r-- 1 snail snail 0 Mar  8 15:46 xonsh
>>> ls = 44
>>> l = 2
>>> ls -l
42
>>> del ls
>>> ls -l
total 0
-rw-rw-r-- 1 snail snail 0 Mar  8 15:46 xonsh
```
This example is official - https://xon.sh/tutorial.html#python-mode-vs-subprocess-mode

So there is no real alternative to Python with clear separation between shell commands and Python syntax in regard to fast system automation.
